### PR TITLE
Fix the issue with reset password when customer has address from not allowed country

### PR DESCRIPTION
**Pull request by dmytro-ch**
### Description
Fix the issue with reset password when customer has an address from not allowed country.

### Fixed Issues (if relevant)
1. magento/magento2#18170: Unable to reset password if customer has address from not allowed country

### Manual testing scenarios
1. Create customer on frontend, add some customer address with Germany (just an example) and log out
2. Go to Admin >> Stores >> Configuration >> General >> General
3. In Country Options >> Allow Countries deselect Germany (mark as not allowed), click to Save Config (probably flushing cache also required)
4. On frontend got to "Forgot password" form and try to reset password for customer that you created in step 1

### Expected result
1. Customer should get reset password link without any issues

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @dmytro-ch. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me $VERSION instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

File: app/code/Magento/Customer/Model/AccountManagement.php [hunk 1]
```diff
@@ -387,7 +393,8 @@ public function __construct(
         AccountConfirmation $accountConfirmation = null,
         SessionManagerInterface $sessionManager = null,
         SaveHandlerInterface $saveHandler = null,
-        CollectionFactory $visitorCollectionFactory = null
+        CollectionFactory $visitorCollectionFactory = null,
+        AddressRegistry $addressRegistry
```
Comment (ishakhsuvarov) [inline app/code/Magento/Customer/Model/AccountManagement.php]: Required parameter can not be placed after optional parameters

Comment (dmytro-ch) [reply app/code/Magento/Customer/Model/AccountManagement.php]: Thank you, @ishakhsuvarov!
I've added the missung default value for the optional parameter.

**Comment by rogyar** [comment]
Please, consider fixing the failing static tests. The issues might not be related directly to your changes but to files that were changed in general. If you need help with that, feel free to connect me in Slack

**Comment by dmytro-ch** [comment]
@rogyar, thank you!
I've fixed the code sniffer and mess detector issues.

**Comment by slavvka** [comment]
@dmytro-ch please look at Codacy review and fix the issue.

**Comment by dmytro-ch** [comment]
@slavvka, thank you! I've fixed the issue mentioned in Codacy review.

**Comment by rogyar** [comment]
@magento-engcom-team give me 2.2.5 instance

**Comment by magento-engcom-team** [comment]
Hi @rogyar. Thank you for your request. I'm working on Magento 2.2.5 instance for you

**Comment by magento-engcom-team** [comment]
Hi @rogyar, here is your [Magento instance](https://i-18179-2-2-5.engcom.dev.magento.com).
Admin access: [https://i-18179-2-2-5.engcom.dev.magento.com/admin](https://i-18179-2-2-5.engcom.dev.magento.com/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by slavvka** [comment]
@dmytro-ch I didn't notice that `$visitorCollectionFactory` existed before. Please could you revert the change back to maintain backward compatibility? You may ignore Codacy fail then. Thank you!

**Comment by dmytro-ch** [comment]
@slavvka, sure, I reverted the changes :)
Thank you.

**Review by slavvka — APPROVED** [review]
Thank you!

**Comment by magento-engcom-team** [comment]
Hi @slavvka, thank you for the review.
[ENGCOM-3056](https://jira.corp.magento.com/browse/ENGCOM-3056) has been created to process this Pull Request

**Comment by slavvka** [comment]
Hey @dmytro-ch during testing we found the problem:

**Problem**: Cannot set New Password using reset password link.

Manual testing scenario:
1. Create customer on frontend.
1. Add customer address with Germany country
1. Log out.
1. Go to Admin >> Stores >> Configuration >> General >> General
1. In Country Options >> Allow Countries deselect Germany (mark as not allowed), click to Save Config.
1. On frontend go to "Forgot password" page.
1. Fill form and click Reset My Password button.
1. Click Set a New Password link from email.
1. Enter new password and click Set a New Password
1. :x: Error message "Invalid value of "DE" provided for the countryId field."
image

**Comment by slavvka** [comment]
Also please resolve the conflicts with mainline

**Comment by dmytro-ch** [comment]
Hello @slavvka,
I've resolved the merge conflict and fixed the mentioned issue.

Thank you!

**Comment by magento-engcom-team** [comment]
Hi @dmytro-ch. Thank you for your contribution. 
We will aim to release these changes as part of 2.3.1. 
Please check the release notes for final confirmation.

**Pull request merged**